[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section

simpal01 opened the pull request:
When the .eh_frame section is placed at a non-zero
offset within its output section, the relocation
value within .eh_frame are computed incorrectly.

We had similar issue in .ARM.exidx section and it has been
fixed already in https://reviews.llvm.org/D148033.

While applying the relocation using S+A-P,  the value
 of P (the location of the relocation) is getting wrong. 
P is:
  P = SecAddr + rel.offset, But SecAddr points to the
starting address of the outputsection rather than the
starting address of the eh frame section within that
output section.

This issue affects all targets which generates .eh_frame 
section. Hence fixing in all the corresponding targets it affecting.

simpal01 committed:
[LLD][AARCH64] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.

When the .eh_frame section is placed at a non-zero
offset within its output section, the relocation
value within .eh_frame are computed incorrectly.

We had similar issue in AArch32 and it has been
fixed already in https://reviews.llvm.org/D148033.

While applying the relocation using S+A-P, the value
of P (the location of the relocation) is getting wrong.
P is:
  P = SecAddr + rel.offset, But SecAddr points to the
starting address of the outputsection rather than the
starting address of the eh frame section within that
output section.

smithp35 reviewed (commented):
This will only fix the problem for AArch64, but it also affects other targets. Assuming relocateAlloc is the best place to fix this, then it would need applying in all instances of relocateAlloc.

simpal01 commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -770,6 +770,9 @@ void AArch64::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
   uint64_t secAddr = sec.getOutputSection()->addr;
   if (auto *s = dyn_cast<InputSection>(&sec))
     secAddr += s->outSecOff;
+  else if (auto *eh = dyn_cast<EhInputSection>(&sec))
```
yeah this affects all other targets (might be X86 and PPC as well).

smithp35 replied:
OK, reading through the code. The .ARM.exidx sections were also InputSections so we could alter their outSecOff to compensate. However EhInputSections are not InputSections so this won't work here. We need to find the outSecOff of the ehframe section (which is usually 0, but with some linker scripts isn't).

I don't know of a legal EhInputSection that would return getParent() == nullptr so it should be possible to remove the nested if.

I suggest use ehIn and ehFrame to make it a bit clearer which sections we are referring to. I've use SyntheticSection as it is what EhInputSection getParent() although that is derived from InputSection so in the code below InputSection would work too.
```
if (auto *ehIn = dyn_cast<EhInputSection>(&sec) {
  SyntheticSection *ehFrame = ehIn->getParent();
  SecAddr += ehFrame->outSecOff; 
}
```

smithp35 reviewed (commented):
Thanks for fixing the commits, that makes it easier to review.

I can see that this would need a fix in relocateAlloc. I think you'll need to update the generic one in Target.cpp as well as the Target specific ones for x86_64 and PPC (and any other I've missed).

You'll need to add x86_64, PPC and at least one other target's (for the generic Target) so that we've covered all the code-paths. By replacing the nops with a common data directive you might be able to make the same file work for all 3. However I think it would still need separate files so that the REQUIRES only had one target.

simpal01 committed:
fixup! [LLD][AARCH64] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.

simpal01 commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -583,9 +583,14 @@ static uint64_t readFdeAddr(uint8_t *buf, int size) {
+  // Adding outSecOff as finalizeAddressDependentContent()
+  // may have altered the corresponding outSecOff. This is
+  // required to get the correct PC relative offset.
+  off = off + outSecOff;
```
on  a second thought, adding outSecOff inside the condition might be  more sensible.
  if ((enc & 0x70) == DW_EH_PE_pcrel)
    return addr + getParent()->addr + off;

smithp35 reviewed (commented):
Can you rename the PR and the commit message to remove [AARCH64] as this is now a general change.

I think with this and your own suggestion of moving the change in getFdePc inside the if statement it will be ready for my approval. We'll want to get MaskRay to take a look too.

simpal01 committed:
fixup! [LLD]lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.

Adding outSecOff in eh_frame_hdr section as
finalizeAddressDependentContent() may have altered
the corresponding outSecOff. This is required to get
the correct initial location of each FDE entries.

simpal01 commented:
> Can you rename the PR and the commit message to remove [AARCH64] as this is now a general change.
> 
> I think with this and your own suggestion of moving the change in getFdePc inside the if statement it will be ready for my approval. We'll want to get MaskRay to take a look too.

Done

simpal01 commented:
@MaskRay Could you please have a look into this patch?

MaskRay commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -770,6 +770,10 @@ void AArch64::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
   if (auto *s = dyn_cast<InputSection>(&sec))
     secAddr += s->outSecOff;
+  else if (auto *ehIn = dyn_cast<EhInputSection>(&sec)) {
+    SyntheticSection *ehFrame = ehIn->getParent();
```
`ehFrame` is used only once. Just inline its use and drop the braces. Ditto everywhere.

MaskRay commented on lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:
```diff
@@ -0,0 +1,55 @@
+// REQUIRES: aarch64
+// RUN: rm -rf %t && split-file %s %t
```
`// RUN: rm -rf %t && split-file %s %t && cd %t` then you can omit `%t/` below.

MaskRay commented on lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:
```diff
@@ -0,0 +1,55 @@
+  .text : { *(.text .text.*) }
+  .eh_frame : {
+  /* Alignment padding within .eh_frame */
+  . = ALIGN(128);
```
Indent input section descriptions within `.eh_frame`.

`. = ALIGN(128);` may not necessarily add a padding if `.` is already aligned.
Use `. += 128;` instead?

simpal01 replied:
Yeah looks like that is the case. Replaced **. = ALIGN(128**) with **. += 128**.

MaskRay commented on lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:
```diff
@@ -0,0 +1,55 @@
+  .text : { *(.text .text.*) }
+  .eh_frame : ALIGN(128) {
+  __eh_frame_start = .;
+  *(.eh_frame .eh_frame.*) ;
```
For the purpose of this test, you can omit `.eh_frame.*`. GNU ld's internal linker script still uses `.eh_frame.*` but it is not clear which port still uses `.eh_frame.*`.

simpal01 replied:
Done

MaskRay commented:
> [LLD] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.

The convention is to use an imperative sentence and omit the trailing dot.

Perhaps `[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section` will be better.
(Some patches use `[LLD][ELF]` or `[lld][ELF]` but I find `[lld]` redundant.)

simpal01 committed:
fixup! [ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section

simpal01 commented:
> > [LLD] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.
> 
> The convention is to use an imperative sentence and omit the trailing dot.
> 
> Perhaps `[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section` will be better. (Some patches use `[LLD][ELF]` or `[lld][ELF]` but I find `[lld]` redundant.)

Done
